integration_tests/test_downsample_mask.py

```diff
@@ -0,0 +1,219 @@
+import os
+import renderapi
+import json
+from test_data import (
+        TEST_DATA_ROOT,
+        render_params)
+
+from rendermodules.rough_align.downsample_mask_handler import (
+        DownsampleMaskHandler, points_in_mask)
+from rendermodules.module.render_module import \
+        RenderModuleException
+import pytest
+import numpy as np
+
+DS_MASK_TEST_PATH = os.path.join(
+        TEST_DATA_ROOT,
+        "em_modules_test_data",
+        "downsample_masking")
+DS_MASK_PATH = os.path.join(
+        DS_MASK_TEST_PATH,
+        "masks")
+DS_BAD_MASK_PATH = os.path.join(
+        DS_MASK_TEST_PATH,
+        "bad_masks")
+DS_TSPEC_PATH = os.path.join(
+        DS_MASK_TEST_PATH,
+        "ds_tilespecs.json")
+DS_COLLECTION_PATH = os.path.join(
+        DS_MASK_TEST_PATH,
+        "ds_matches.json")
+
+
+@pytest.fixture(scope='module')
+def render():
+    render_params['project'] = 'downsample_mask_test'
+    render = renderapi.connect(**render_params)
+    return render
+
+
+@pytest.fixture(scope='module')
+def downsampled_stack(render):
+    with open(DS_TSPEC_PATH, 'r') as f:
+        tspecs = [
+            renderapi.tilespec.TileSpec(json=i)
+            for i in json.load(f)]
+
+    stack = "downsampled_stack_for_masks"
+    renderapi.stack.create_stack(stack, render=render)
+    renderapi.client.import_tilespecs(stack,
+                                      tspecs,
+                                      render=render)
+    renderapi.stack.set_stack_state(stack,
+                                    'COMPLETE',
+                                    render=render)
+    yield stack
+
+
+@pytest.fixture(scope='module')
+def downsampled_collection(render):
+    collection = "downsampled_collection_for_masks"
+    with open(DS_COLLECTION_PATH, 'r') as f:
+        renderapi.pointmatch.import_matches(
+                collection,
+                json.load(f),
+                render=render)
+    yield collection
+
+
+def zs_have_masks(render, stack):
+    tspecs = renderapi.tilespec.get_tile_specs_from_stack(
+            stack,
+            render=render)
+    zs = [{'z': t.z, 'group': t.layout.sectionId}
+          for t in tspecs if t.ip[0].maskUrl is not None]
+    return zs
+
+
+def matches_are_filtered(render, collection):
+    groups = renderapi.pointmatch.get_match_groupIds(
+            collection,
+            render=render)
+    matches = []
+    for i in range(len(groups)):
+        for j in range(i):
```

Comment:
This is probably more readable as `for group1, group2 in itertools.combinations(groups, 2):`, which is ordered differently but shouldn't matter in the renderapi call.

Reply:
agreed